stage0/src/alloc.rs

```diff
@@ -0,0 +1,125 @@
+
+struct Inner<const N: usize> {
+    index: AtomicUsize,
+    storage: MaybeUninit<[u8; N]>,
+}
+
+impl<const N: usize> Inner<N> {
+    const fn new() -> Self {
+        Self {
+            index: AtomicUsize::new(0),
+            storage: MaybeUninit::uninit(),
+        }
+    }
+}
+
+/// Basic bump allocator that never deallocates memory.
+pub struct Allocator<const N: usize> {
+    inner: Spinlock<Inner<N>>,
+}
+
+impl<const N: usize> Allocator<N> {
+    pub const fn uninit() -> Self {
+        Self {
+            inner: Spinlock::new(Inner::new()),
+        }
+    }
+
+    pub fn allocate(&self, layout: Layout) -> Option<NonNull<u8>> {
+        let mut inner = self.inner.lock();
+        let offset = inner.index.load(Ordering::Acquire);
+        let storage_ptr = inner.storage.as_mut_ptr() as *mut u8;
+
+        // We may need to allocate extra bytes to ensure proper alignment in memory.
+        let align = if (storage_ptr as usize + offset) % layout.align() > 0 {
+            layout.align() - ((storage_ptr as usize + offset) % layout.align())
+        } else {
+            0
+        };
+
+        if offset + align + layout.size() > N {
+            // memory exhausted
+            return None;
+        }
+
+        inner
+            .index
+            .fetch_add(align + layout.size(), Ordering::SeqCst);
+
+        // Safety: we've reserved memory from [offset, offset + align + size) so this will not
+        // exceed the bounds of `self.storage` and is not aliased.
+        NonNull::new(unsafe { storage_ptr.add(offset + align) })
+    }
+
+    pub fn leak<T>(&self, val: T) -> Option<&mut T> {
+        let ptr = self.allocate(Layout::for_value(&val))?.cast().as_ptr();
+
+        // Safety: we've successfully allocated enough memory that is of the correct size, therefore
+        // writing `val` to it is safe. We also ensure that the pointer is properly initialized and
+        // aligned, so dereferencing it as a reference is fine.
+        unsafe {
+            write(ptr, val);
+            ptr.as_mut()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn simple_alloc() {
+        let alloc = Allocator::<16>::uninit();
+        let val = alloc.leak([0u8; 16]);
+        assert!(val.is_some());
+        let val = alloc.leak([0u8; 16]);
+        assert!(val.is_none());
+    }
+
+    #[test]
+    fn two_alloc() {
+        let alloc = Allocator::<16>::uninit();
+        let val = alloc.leak([0u8; 8]);
+        assert!(val.is_some());
+        let val = alloc.leak([0u8; 8]);
+        assert!(val.is_some());
+    }
+
+    #[test]
+    fn aligned_alloc() {
+        #[repr(align(16))]
+        struct Foo {
+            x: u8,
+        }
+        let alloc = Allocator::<24>::uninit();
+        let val = alloc.leak(Foo { x: 16 }).unwrap();
+        assert_eq!(0, val as *const Foo as usize % core::mem::align_of::<Foo>());
+        assert_eq!(16, val.x);
+        let val = alloc.leak(Foo { x: 32 });
```

Comment:
I am assuming the intention for this test is to enter line 57 and align something. Am I correct?

If so, could we add something to show how much was aligned in this test, i.e., a second alloc which actually works out, and then one with one more byte which fails?

Reply:
Yes, this is to check that the alignment works.

I've changed the test around a bit, but unfortunately it's difficult to know if we can allocate even one more byte: the padding necessary depends on where exactly the buffer lands in memory, as that doesn't have any particular padding requirements.
Worst case, you'll need 15 bytes of padding (if we were off by one byte by chance) + memory for the data structure itself. This randomness makes it hard to figure out whether we have any memory left over or not.